### PR TITLE
Fix ConsoleExporter fails silently when exporting an `ActivityLink` without Tags.

### LINKED ISSUE #3863: [Console Exporter] Partial output when Activity has a link without any tags due to NullReferenceException
# Bug Report

```xml
<PackageReference Include="OpenTelemetry.Exporter.Console" Version="1.4.0-beta.2" />
<TargetFramework>net6.0</TargetFramework>
```

## Symptom

With an `Activity` that includes an `ActivityLink` without any tags, the console exporter throws a `NullReferenceException` when stopping/disposing the `Activity`.

**What is the expected behavior?**

The console output shows all activity information including the "Resource associated with Activity".

**What is the actual behavior?**

The link is included in the output, but not the "Resource associated with Activity" afterwards. A `NullReferenceException` is thrown, which is caught by the SDK and included in the diagnostics log:

```txt
2022-11-03T15:44:00.6954123Z:Unknown error in SpanProcessor event '{0}': '{1}'.{OnExport}{System.NullReferenceException: Object reference not set to an instance of an object.
   at OpenTelemetry.Exporter.ConsoleActivityExporter.Export(Batch`1& batch)
   at OpenTelemetry.SimpleExportProcessor`1.OnExport(T data)}
```

Visual Studio breaks for the exception by default, but execution can be resumed (this is configurable).

## Reproduce

```csharp
using System.Diagnostics;
using OpenTelemetry;
using OpenTelemetry.Trace;

var source = new ActivitySource("Testing");
using var tracerProvider = Sdk.CreateTracerProviderBuilder()
    .AddSource("Testing")
    .AddConsoleExporter()
    .Build();

ActivityContext context;
using (var first = source.StartActivity("first"))
{
    context = first!.Context;
}

var links = new[] { new ActivityLink(context) };
using var second = source.StartActivity(ActivityKind.Internal, links: links, name: "Second");
```

## Additional Context

The problem is in [ConsoleActivityExporter.cs](https://github.com/open-telemetry/opentelemetry-dotnet/blob/6c3beef649a43f6b995877d04891d764bc72995c/src/OpenTelemetry.Exporter.Console/ConsoleActivityExporter.cs#L119) where it has a `foreach` to enumerate the tags without first checking for `null`. In the same file, it looks like it may also have an issue with events that don't have any tags, but I haven't testing that case. The code should probably switch to using `activityLink.EnumerateTags()` (like in #3580).

### PR DESCRIPTION
**Pull request by TimothyMothra**
Fixes #3863 .

ConsoleExporter fails silently when exporting an `ActivityLink` without Tags.

## Changes
- added null check to `ConsoleActivityExporter`
- added new test class `test/OpenTelemetry.Tests/Trace/ConsoleExporterTest.cs` to reproduce issue
  - added ConsoleExporter project to `OpenTelemetry.Tests.csproj`
- added test helpers to catch Errors logged to EventSource
  - `UnitTestEventListener` to subscribe to specific EventSourceName and EventLevel 
  `public UnitTestEventListener(string eventName, EventLevel eventLevel)`
  - `VerifyNoEventSourceErrorsLoggedTestAttribute` used to declare that a unit test should not generate Errors.

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [ ] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: test/OpenTelemetry.Tests/Trace/ConsoleExporterTest.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+                context = first!.Context;
+            }
+
+            var links = new[] { new ActivityLink(context) };
+            using (var secondActivity = source.StartActivity(ActivityKind.Internal, links: links, name: "Second"))
+            {
+            }
+
+            // Wait for the Activity to dispose and be exported.
+            Assert.True(SpinWait.SpinUntil(
```
Comment (reyang) [inline test/OpenTelemetry.Tests/Trace/ConsoleExporterTest.cs:59]: Why is wait needed?

Comment (TimothyMothra) [reply test/OpenTelemetry.Tests/Trace/ConsoleExporterTest.cs:59]: The test can finish executing before the Activity has been exported.

Comment (reyang) [reply test/OpenTelemetry.Tests/Trace/ConsoleExporterTest.cs:59]: Could you elaborate why?

Comment (TimothyMothra) [reply test/OpenTelemetry.Tests/Trace/ConsoleExporterTest.cs:59]: Don't have a clear understanding on the why, it was reproducing consistently yesterday.
It's possible that I wasn't disposing objects correctly. 
It's not reproducing now so I will remove this.

File: test/OpenTelemetry.Tests/Trace/ConsoleExporterTest.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        /// <summary>
+        /// Test case for https://github.com/open-telemetry/opentelemetry-dotnet/issues/3863.
+        /// </summary>
+        [Fact]
+        public void VerifyConsoleActivityExporterDoesntFailWithoutActivityLinkTags()
+        {
+            var exportedItems = new List<Activity>();
+
+            using var source = new ActivitySource("Testing");
```
Comment (reyang) [inline test/OpenTelemetry.Tests/Trace/ConsoleExporterTest.cs:37]: Use test case specific/unique source name to avoid conflict.

File: test/OpenTelemetry.Tests/Trace/ConsoleExporterTest.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+                context = first!.Context;
+            }
+
+            var links = new[] { new ActivityLink(context) };
+            using (var secondActivity = source.StartActivity(ActivityKind.Internal, links: links, name: "Second"))
+            {
+            }
+
+            // Assert that an Activity was exported where ActivityLink.Tags == null.
+            var activity = exportedItems.First(x => x.DisplayName == "Second");
```
Comment (reyang) [inline test/OpenTelemetry.Tests/Trace/ConsoleExporterTest.cs:57]: Simply the code, Linq can be avoided with proper management of the exported list.

File: test/OpenTelemetry.Tests/Shared/VerifyNoEventSourceErrorsLoggedTestAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    using System.Diagnostics.Tracing;
+    using System.Linq;
+    using System.Reflection;
+    using Xunit;
+    using Xunit.Sdk;
+
+    /// <summary>
+    /// Can be used to assert that no Errors were logged to EventSource during a unit test.
+    /// </summary>
+    internal class VerifyNoEventSourceErrorsLoggedTestAttribute : BeforeAfterTestAttribute
```
Comment (reyang) [inline test/OpenTelemetry.Tests/Shared/VerifyNoEventSourceErrorsLoggedTestAttribute.cs:28]: Is this related to the PR?

**Review by reyang — CHANGES_REQUESTED** [review]
`UnitTestEventListener` is beyond the scope of the issue/PR.

File: src/OpenTelemetry.Exporter.Console/ConsoleActivityExporter.cs [hunk 5]
```diff
@@ -116,11 +116,14 @@ public override ExportResult Export(in Batch<Activity> batch)
                     foreach (var activityLink in activity.Links)
                     {
                         this.WriteLine($"    {activityLink.Context.TraceId} {activityLink.Context.SpanId}");
-                        foreach (var attribute in activityLink.Tags)
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Exporter.Console/ConsoleActivityExporter.cs:119]: @TimothyMothra FYI you could re-write this as:

```csharp
foreach (ref readonly var attribute in activityLink.EnumerateTagObjects())
```

More efficient and it will handle the null case for you 😄

Might be worth re-writing all of the foreach loops in here to use the newer pattern just to have a good reference doing it correctly. Follow-up thing if you are interested 😉

File: test/OpenTelemetry.Tests/OpenTelemetry.Tests.csproj [hunk 6]
```diff
@@ -10,6 +10,7 @@
   <ItemGroup>
     <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry\OpenTelemetry.csproj" />
     <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.InMemory\OpenTelemetry.Exporter.InMemory.csproj" />
+    <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.Console\OpenTelemetry.Exporter.Console.csproj" />
```
[the remark this answers is not in the document]
Comment (TimothyMothra) [reply test/OpenTelemetry.Tests/OpenTelemetry.Tests.csproj:13]: I could go either way on this.
We currently use this project to validate the InMemoryExporter so I don't think one more exporter is a big deal.
But if we want to invest in more tests for ConsoleExporter specifically, then yeah a new test project would make sense. :)

File: test/OpenTelemetry.Tests/Shared/VerifyNoEventSourceErrorsLoggedTestAttribute.cs (hunk 4 above)
Comment (TimothyMothra) [reply test/OpenTelemetry.Tests/Shared/VerifyNoEventSourceErrorsLoggedTestAttribute.cs:28]: ConsoleExporter was failing silently. I was looking for a way to capture Errors logged to EventSource as that's our only indication that something went wrong. 
I think I can rewrite the test to catch the Exporter's exception now.

Longer term, I think there's value in having a mechanism to detect any other silent errors in the SDK.

File: test/OpenTelemetry.Exporter.Console.Tests/ConsoleActivityExporterTest.cs [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        // Assert that ConsoleExporter does not throw.
+        try
+        {
+            var consoleExporter = new ConsoleActivityExporter(new ConsoleExporterOptions());
+
+            consoleExporter.Export(new Batch<Activity>(activity));
+        }
+        catch (Exception ex)
+        {
+            Assert.Fail(ex.Message);
```
Comment (reyang) [inline test/OpenTelemetry.Exporter.Console.Tests/ConsoleActivityExporterTest.cs:72]: This seems weird, why is it needed?

File: test/OpenTelemetry.Tests/OpenTelemetry.Tests.csproj (hunk 6 above)
[the remark this answers is not in the document]
Comment (TimothyMothra) [reply test/OpenTelemetry.Tests/OpenTelemetry.Tests.csproj]: @CodeBlanch should the new project be a separate PR? 
Looks like there's some side effects in the SLN (alphabetizing items, changing guids).

File: test/OpenTelemetry.Exporter.Console.Tests/ConsoleActivityExporterTest.cs [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+public class ConsoleActivityExporterTest
+{
+    /// <summary>
+    /// Test case for https://github.com/open-telemetry/opentelemetry-dotnet/issues/3863.
+    /// </summary>
+    [Fact]
+    public void VerifyConsoleActivityExporterDoesntFailWithoutActivityLinkTags()
+    {
+        var uniqueTestId = Guid.NewGuid();
+        var activitySourceName = $"activitySourceName{uniqueTestId}";
```
Comment (reyang) [inline test/OpenTelemetry.Exporter.Console.Tests/ConsoleActivityExporterTest.cs:35]: What's the common practice?

File: test/OpenTelemetry.Exporter.Console.Tests/ConsoleActivityExporterTest.cs [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+        // Assert that an Activity was exported where ActivityLink.Tags == null.
+        var activity = exportedItems[0];
+        Assert.Equal("Second", activity.DisplayName);
+        Assert.Null(activity.Links.First().Tags);
+
+        // Assert that ConsoleExporter does not throw.
+        try
+        {
+            var consoleExporter = new ConsoleActivityExporter(new ConsoleExporterOptions());
```
Comment (reyang) [inline test/OpenTelemetry.Exporter.Console.Tests/ConsoleActivityExporterTest.cs:66]: Lifecycle management?

File: test/OpenTelemetry.Tests/OpenTelemetry.Tests.csproj (hunk 6 above)
[the remark this answers is not in the document]
Comment (CodeBlanch) [reply test/OpenTelemetry.Tests/OpenTelemetry.Tests.csproj]: I'm good with doing as a follow-up.

File: test/OpenTelemetry.Exporter.Console.Tests/ConsoleActivityExporterTest.cs (hunk 7 above)
Comment (TimothyMothra) [reply test/OpenTelemetry.Exporter.Console.Tests/ConsoleActivityExporterTest.cs:72]: My instinct is to assert any expected behavior in the test. But you're right, this isn't necessary as the throw Exception will fail the test. I'll remove this.

File: test/OpenTelemetry.Exporter.Console.Tests/ConsoleActivityExporterTest.cs (hunk 8 above)
Comment (reyang) [reply test/OpenTelemetry.Exporter.Console.Tests/ConsoleActivityExporterTest.cs:35]: https://github.com/open-telemetry/opentelemetry-dotnet/blob/943c856c1c6295a536bc8a1c89896febc0be6107/test/OpenTelemetry.Exporter.Prometheus.HttpListener.Tests/PrometheusSerializerTests.cs#L34

Comment (TimothyMothra) [reply test/OpenTelemetry.Exporter.Console.Tests/ConsoleActivityExporterTest.cs:35]: There's not a clear common practice.
In this repo I can see nameof(TestMethod), nameof(TestClass), and arbitrary strings used.
The Guid will give the strongest guarantee against duplicates and is safe from future copy/paste reusing this code block.

Comment (reyang) [reply test/OpenTelemetry.Exporter.Console.Tests/ConsoleActivityExporterTest.cs:35]: Do you think it makes sense to ask for convergence instead of creating yet another way?

Comment (reyang) [reply test/OpenTelemetry.Exporter.Console.Tests/ConsoleActivityExporterTest.cs:35]: > The Guid will give the strongest guarantee against duplicates and is safe from future copy/paste reusing this code block.

Could you see what's the downside?

File: test/OpenTelemetry.Tests/OpenTelemetry.Tests.csproj (hunk 6 above)
[the remark this answers is not in the document]
Comment (TimothyMothra) [reply test/OpenTelemetry.Tests/OpenTelemetry.Tests.csproj]: #3933

File: test/OpenTelemetry.Exporter.Console.Tests/ConsoleActivityExporterTest.cs (hunk 8 above)
Comment (reyang) [reply test/OpenTelemetry.Exporter.Console.Tests/ConsoleActivityExporterTest.cs]: TBH I haven't spent time on this, some negative outcomes I can see are that a randomly generated GUID could be 1) very random, making it hard to replay 2) not following the normal naming convention, making it anti-pattern

Comment (reyang) [reply test/OpenTelemetry.Exporter.Console.Tests/ConsoleActivityExporterTest.cs]: My suggestion is that if you want to take an approach, think through the pros/cons.

**Comment by TimothyMothra** [comment]
I'm waiting for #3933 and then i'll rebase this PR.

File: test/OpenTelemetry.Exporter.Console.Tests/ConsoleActivityExporterTest.cs [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    public void VerifyConsoleActivityExporterDoesntFailWithoutActivityLinkTags()
+    {
+        var activitySourceName = Utils.GetCurrentMethodName();
+        using var activitySource = new ActivitySource(activitySourceName);
+
+        var exportedItems = new List<Activity>();
+
+        using var tracerProvider = Sdk.CreateTracerProviderBuilder()
+            .AddSource(activitySourceName)
+            .AddInMemoryExporter(exportedItems)
```
Comment (utpilla) [inline test/OpenTelemetry.Exporter.Console.Tests/ConsoleActivityExporterTest.cs:36]: This could be simplified by just using the extension method `AddConsoleExporter`. We don't have to test if activities are actually created in this test.

Comment (TimothyMothra) [reply test/OpenTelemetry.Exporter.Console.Tests/ConsoleActivityExporterTest.cs:36]: The purpose of the InMemoryExporter was to inspect the Activity that was created and verify that it doesn't have Tags. This is the specific condition that was failing.

Comment (utpilla) [reply test/OpenTelemetry.Exporter.Console.Tests/ConsoleActivityExporterTest.cs:36]: Ahh I remember this from our discussion now. It's caught in the SimpleExportProcessor.

**Review by utpilla — APPROVED** [review]

File: src/OpenTelemetry.Exporter.Console/ConsoleActivityExporter.cs (hunk 5 above)
Comment (TimothyMothra) [reply src/OpenTelemetry.Exporter.Console/ConsoleActivityExporter.cs:119]: I'll take this as a follow up after this PR merges. :)

**Review by Kielek — APPROVED** [review]

File: src/OpenTelemetry/AssemblyInfo.cs [hunk 11]
```diff
@@ -17,6 +17,7 @@
 using System.Runtime.CompilerServices;
 
 [assembly: InternalsVisibleTo("OpenTelemetry.Tests" + AssemblyInfo.PublicKey)]
+[assembly: InternalsVisibleTo("OpenTelemetry.Exporter.Console.Tests" + AssemblyInfo.PublicKey)]
```
Comment (utpilla) [inline src/OpenTelemetry/AssemblyInfo.cs:20]: What `internal` member of the ConsoleExporter do the tests need access to?

Comment (TimothyMothra) [reply src/OpenTelemetry/AssemblyInfo.cs]: The test was using an internal ctor for `Batch`. 
This test was originally in another project (with internals) before moving to its own project.

I made a change to use a public ctor and removed this.

File: src/OpenTelemetry.Exporter.Console/CHANGELOG.md [hunk 12]
```diff
@@ -2,6 +2,10 @@
 
 ## Unreleased
 
+* Bug fix, ConsoleExporter fails silently when exporting an `ActivityLink`
```
Comment (utpilla) [inline src/OpenTelemetry.Exporter.Console/CHANGELOG.md:5]: nit: Reword this to something like:
Bug fix to prevent ConsoleExporter from failing when exporting an `ActivityLink` without tags.

Comment (utpilla) [reply src/OpenTelemetry.Exporter.Console/CHANGELOG.md:5]: Not a blocker.

Comment (TimothyMothra) [reply src/OpenTelemetry.Exporter.Console/CHANGELOG.md:5]: https://github.com/open-telemetry/opentelemetry-dotnet/pull/3984
